### PR TITLE
filestate/projectStore: Optimize ListReferences

### PR DESCRIPTION
**Pull request by abhinav**
ListReferences for project layout is currently inefficient.
It roughly does the equivalent of:

    for proj in bucket.List(stacksDir)
        for ref in bucket.List(stacksDir + "/" + proj):
            # ...

This is fine for local, file-based storage,
but super inefficient for S3 or Azure.
The reason this was written as above is because we treat the bucket
as a directory tree in listBucket.

This changes projectStore.ListReferences to use the underlying
bucket.List method directly--without setting the Delimiter.
This treats the bucket as a flat file list instead of a hierarchy,
and we can extract project and stack names from it in one go.

### CODE REVIEW

**Comment by abhinav** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12485** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/12485" target="_blank">Graphite</a> 
    * **PR #12437** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/12437" target="_blank">Graphite</a> 
      * **PR #12438** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/12438" target="_blank">Graphite</a> 
        * **PR #12496** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/12496" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/pulumi/pulumi/12496?utm_source=stack-comment).

File: pkg/backend/filestate/store.go [hunk 1]
```diff
@@ -232,49 +235,63 @@ func (p *projectReferenceStore) ListProjects() ([]tokens.Name, error) {
...
+		// Key is in the form,
+		//   $StacksDir/$projName/$stackName.json[.gz]
+		// We want to extract projName and stackName from it.
 
-			if _, has := encoding.Marshalers[ext]; !has {
-				continue
-			}
+		parts := strings.Split(strings.TrimPrefix(file.Key, prefix), "/")
+		if len(parts) != 2 {
+			continue // skip paths too shallow or too deep
```
Comment (dixler) [inline pkg/backend/filestate/store.go:267]: Nit: maybe comment something of the nature

```suggestion
			continue // skip paths that do not match the specification
```

File: pkg/backend/filestate/store.go [hunk 2]
```diff
@@ -232,49 +235,63 @@ func (p *projectReferenceStore) ListProjects() ([]tokens.Name, error) {
...
 
 	var stacks []*localBackendReference
-	for _, projName := range projects {
-		// TODO: Could we improve the efficiency here by firstly making listBucket return an enumerator not
-		// eagerly collecting all keys into a slice, and secondly by getting listBucket to return all
-		// descendent items not just the immediate children. We could then do the necessary splitting by
-		// file paths here to work out project names.
-		projectFiles, err := listBucket(p.bucket, filepath.Join(path, projName.String()))
+	for {
+		file, err := iter.Next(ctx)
```
Comment (dixler) [inline pkg/backend/filestate/store.go:249]: Is this fast on remote buckets?

**Review by dixler — COMMENTED** [review]
Some initial questions.

**Review by dixler — APPROVED** [review]

File: pkg/backend/filestate/store.go (hunk 2 above)
Comment (Frassle) [reply pkg/backend/filestate/store.go:249]: Should be, means gocloud doesn't _have_ to load the whole bucket at once.

**Review by Frassle — APPROVED** [review]

File: pkg/backend/filestate/store.go (hunk 2 above)
Comment (abhinav) [reply pkg/backend/filestate/store.go:249]: It'll use whatever the underlying pagination is.
The driver.Bucket interface only has a [ListPaged method](https://pkg.go.dev/gocloud.dev@v0.29.0/blob/driver#Bucket.ListPaged), so the way this'll work is S3 will give you the first, say, 100 items, then Next will iterate through them in memory until it runs out and needs the next 100.

And the "flatten" behavior is the bucket's default: S3 doesn't have a concept of directories beyond buckets, so "foo/bar" is just a key; "/" has no meaning to S3 if you don't set the delimiter.

**Comment by abhinav** [comment]
bors merge

**Comment by abhinav** [comment]
Ugh, the merge will probably time out
image

**Comment by abhinav** [comment]
bors cancel

**Comment by abhinav** [comment]
bors merge

**Comment by abhinav** [comment]
bors cancel

**Comment by abhinav** [comment]
bors merge

**Pull request merged**